sycl/doc/design/DeviceConfigFile.md

```diff
@@ -0,0 +1,297 @@
+struct TargetInfo {
+  bool maySupportOtherAspects;
+  std::vector<unsigned> aspects;
+  std::vector<unsigned> subGroupSizes;
+  std::string aotToolchain;
+  std::string oclocDevice;
+};
+
+#include "device_config_file.inc"
+}; // namespace DeviceConfigFile
+```
+
+Other modules can query the map to get the information like in the example
+below:
+```c++ 
+DeviceConfigFile::TargetInfo info = DeviceConfigFile::targets.find("TargetA");
+if (info == DeviceConfigFile::targets.end()) {
+  /* Target not found */
+  ...
+} else {
+  auto aspects = info.aspects;
+  auto maySupportOtherAspects = info.maySupportOtherAspects;
+  auto subGroupSizes = info.subGroupSizes;
+  ...
+}
+```
+
+### Changes to Build Infrastructure
+We need the information about the targets in multiple tools and compiler
+modules listed in [Requirements](#Requirements).  Thus, we need to make sure
+that the generation of the `.inc` file out of the `.td` file is done in time
+for all the consumers. The command we need to run for TableGen is `llvm-tblgen
+-gen-dynamic-tables -I /llvm-root/llvm/include/ input.td -o output.inc`.
+Additionally, we need to set dependencies adequately so that this command is
+run before any of the consumers need it.
+
+### Changes to the DPC++ Frontend
+To allow users to add new targets we provide a new flag:
+`fsycl-device-config-file=/path/to/file.yaml`. Users can pass a `.yaml` file
+describing the targets to be added/updated. An example of how such .yaml` file
+should look like is shown below.
+```
+intel_gpu_skl:
+    aspects: [aspect_name1, aspect_name2]
+    may_support_other_aspects: true/false
+    sub-group-sizes: [1, 2, 4, 8]
+    aot-toolchain: ocloc
+    ocloc-device: skl
+```
+The frontend module should parse the user-provided `.yaml` file and update the
+map with the new information about targets. LLVM provides
+[YAML/IO](https://llvm.org/docs/YamlIO.html) library to easily parse `.yaml`
+files.
+
+As mentioned in [Requirements](#Requirements), there is an auto-detection
+mechanism for `aot-toolchain` and `ocloc-device` inferring these from the target
+name. In the `.yaml` example shown above the target name is `intel_gpu_skl`.
+From that name, we can infer that `aot-toolchain` is `ocloc` because the name
+starts with `intel_gpu`. Also, we can infer that `ocloc-device` is `skl` just by
+keeping what is left after the prefix `intel_gpu`.
+
+#### Auto-detection Potential Issues/Limitations
+The auto-detection mechanism is a best effort to relieve users from specifying
+`aot-toolchain` and `ocloc-device` from well known devices. However, it has
+limitations and potential issues:
+- Rules for target names: **TODO: Define rules for names so that they can be
+auto-detected.**
+- Multiple targets with the same name: On the one hand, the compiler emits a
+warning so that the user is aware that multiple targets share the same name. On
+the other hand, it simply processes each new entry and updates the map with the
+latest information found.
```

Comment:
I don't understand the issue with multiple targets.  Is this specific to auto-detection?  Or is it a general problem that the user could specify the same target name multiple times in the YAML file?

Reply:
Yes, I agree that it is probably an error, and you can raise a diagnostic in this case.

It's a little confusing to describe this in the section titled "auto-detection potential issues" because it is unrelated to auto-detection.

Reply:
.> It's a little confusing to describe this in the section titled "auto-detection potential issues" because it is unrelated to auto-detection.

That's a fair point, I think I will change it to "Potential Issues".